exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,490 @@
+    ///     fn verify(&self) -> bool {
+    ///         self.verify_signature(self.public_key())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn execute(&self, _: &mut Fork) -> ExecutionResult {
+    /// #       Ok(())
+    /// #   }
+    /// }
+    /// # fn main() {}
+    fn verify(&self) -> bool;
+
+    /// Receives a fork of the current blockchain state and can modify it depending on the contents
+    /// of the transaction.
+    ///
+    /// # Notes
+    ///
+    /// - Transaction itself is considered committed regardless whether `Ok` or `Err` has been
+    ///   returned or even if panic occurred during execution.
+    /// - Changes made by the transaction are discarded if `Err` is returned or panic occurred.
+    /// - A transaction execution status (see `ExecutionResult` and `TransactionResult` for the
+    ///   details) is stored in the blockchain and can be accessed through api.
+    /// - Blockchain state hash is affected by the transactions execution status.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, ExecutionResult};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #
+    /// #       public_key: &PublicKey,
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Return execution status.
+    ///         Ok(())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult;
+}
+
+/// Result of unsuccessful transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ExecutionError {
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    code: u8,
+    /// Optional error description.
+    description: Option<String>,
+}
+
+impl ExecutionError {
+    /// Constructs a new `ExecutionError` instance with the given error code.
+    pub fn new(code: u8) -> Self {
+        Self {
+            code,
+            description: None,
+        }
+    }
+
+    /// Constructs a new `ExecutionError` instance with the given error code and description.
+    pub fn with_description(code: u8, description: String) -> Self {
+        Self {
+            code,
+            description: Some(description),
+        }
+    }
+}
+
+/// Type of the transaction error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionErrorType {
```

Comment:
I don't quite like such a long name for a type, but can't think of a better alternative rn. Probably, something like
```rust
pub enum ErrorCode {
    Panic,
    User(u8),
}
```

but the `User` variant name is surely not the best one.

Reply:
I don't like that name, either, but it should not be often used directly, especially in the user code.